### PR TITLE
In 150411707 include selected people in contact post

### PR DESCRIPTION
**Pull request by saraOSI**
### Pivotal Story

- [Investigator can select who they talked to during a contact #150411707](https://www.pivotaltracker.com/story/show/150411707)

### Purpose
A user can select the people he/she talked to for a contact

### Notes for Reviewer
There are acceptance criteria for validation on save for people. I'll add that in as a separate pull request, since @NateOSI is still working on a chunk of that code. Also, there wasn't any acceptance criteria for seeing the selected people on the show page. Checking in with product and will add that as a follow-up PR if needed.

### CODE REVIEW

File: app/javascript/investigations/ContactForm.jsx [hunk 1]
```diff
@@ -92,10 +93,14 @@ class ContactForm extends React.Component {
                 }
                 <div className='row'>
                   <FormField gridClassName='col-md-12' label='People Present' id='people'>
-                    { <ul>
-                      {people.map((person, i) => <li key={i}>{person.name}</li>)}
-                    </ul>
-                    }
+                    { people.map((person, i) =>
+                      <CheckboxField
```
Comment (NateOSI) [inline app/javascript/investigations/ContactForm.jsx]: index or i?

File: app/javascript/investigations/ContactForm.jsx [hunk 2]
```diff
@@ -169,6 +174,7 @@ ContactForm.propTypes = {
 
 ContactForm.defaultProps = {
   errors: {},
+  people: [],
```
Comment (NateOSI) [inline app/javascript/investigations/ContactForm.jsx]: why is this a default property?

File: app/javascript/investigations/ContactFormContainer.jsx [hunk 3]
```diff
@@ -45,7 +45,7 @@ const mapStateToProps = (state, ownProps) => {
     communicationMethods: getCommunicationMethodsSelector(state).toJS(),
     inPersonCode: getInPersonCommunicationMethodValueSelector(state),
     locations: getLocationsSelector(state).toJS(),
-    people: contactForm.get('people').toJS(),
+    people: contactForm.get('people') && contactForm.get('people').toJS(),
```
Comment (NateOSI) [inline app/javascript/investigations/ContactFormContainer.jsx]: should this be a selector?

File: app/javascript/investigations/ContactForm.jsx [hunk 4]
```diff
@@ -99,7 +99,13 @@ class ContactForm extends React.Component {
                         id={`person_${i}`}
                         value={person.name}
                         checked={person.selected}
-                        onChange={() => selectPerson(i)}
+                        onChange={(e) => {
```
Comment (NateOSI) [inline app/javascript/investigations/ContactForm.jsx]: you could use deconstructive assignment here in the params ({target: {checked}})

File: app/javascript/investigations/ContactFormContainer.jsx (hunk 3 above)
Comment (saraOSI) [reply app/javascript/investigations/ContactFormContainer.jsx]: It should indeed. I made it one in a later commit.

File: app/javascript/investigations/ContactForm.jsx (hunk 2 above)
Comment (saraOSI) [reply app/javascript/investigations/ContactForm.jsx]: I forgot to remove this when I added a selector that returns a default. Will do so now. Same with the `selectedPeopleIds`.

File: app/javascript/investigations/ContactForm.jsx (hunk 1 above)
Comment (saraOSI) [reply app/javascript/investigations/ContactForm.jsx]: Index is probably better. Will update.

File: app/javascript/investigations/ContactForm.jsx (hunk 4 above)
Comment (saraOSI) [reply app/javascript/investigations/ContactForm.jsx]: I did not think of that, but I like it!

File: app/javascript/reducers/contactFormReducer.js [hunk 5]
```diff
@@ -42,6 +43,7 @@ export default createReducer(Map(), {
       investigation_started_at: {
         value: investigation_started_at,
       },
+      people: people.map((person) => ({name: nameFormatter(person), selected: false, id: person.legacy_descriptor})),
```
[the remark this answers is not in the document]
Comment (saraOSI) [reply app/javascript/reducers/contactFormReducer.js]: That would make sense to me if there was a reason to maintain a copy of the person object as it was returned from the API, as well as have a version with the formatted name. The only piece we need to communicate back/forth with the API is the legacy_descriptor; the various pieces of the full name are somewhat irrelevant to the form, and I wanted to keep the structure of this object limited to what we're actually using in the form.

All of that having been said, I'm open to talking about it some more.

File: spec/features/contacts/create_investigation_contact_spec.rb [hunk 6]
```diff
@@ -25,13 +25,14 @@
...
       select 'School', from: 'Location'
+      find('label', text: 'Emma Woodhouse').click
       select 'Investigate Referral', from: 'Purpose'
       fill_in 'Contact Notes', with: 'This was an attempted contact'
     end
     expect(page).to have_field('Date/Time', with: '08/17/2016 3:00 AM')
     expect(page).to have_select('Communication Method', selected: 'In person')
-    expect(page).to have_content('Emma Woodhouse')
-    expect(page).to have_content('George Knightley')
+    expect(find(:checkbox, 'Emma Woodhouse')).to be_checked
```
[the remark this answers is not in the document]
Comment (saraOSI) [reply spec/features/contacts/create_investigation_contact_spec.rb]: Sure didn't. I'll take a peek at those.

File: app/javascript/reducers/contactFormReducer.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (saraOSI) [reply app/javascript/reducers/contactFormReducer.js]: Nate and I chatted about this some more, and we decided to use a selector to do the presentation for the names. We found a reasonable use case in which we would want the original data structure intact: sorting. That's very likely to happen in a future story, so we want to make sure that the original data structure is preserved enough to make sorting easy. That, combined with making sure the reducer doesn't do too much work for view concerns, led us to agree that formatting the name belongs in a selector.

File: spec/javascripts/reducers/contactFormReducerSpec.js [hunk 7]
```diff
@@ -13,7 +15,11 @@ describe('contactReducer', () => {
 
   describe('on BUILD_CONTACT_SUCCESS', () => {
     it('returns the contact', () => {
-      const action = buildSuccess({investigation_id: '123', investigation_started_at: 'date time'})
+      const people = [
+        {first_name: 'Bob', last_name: 'Smith', legacy_descriptor: '1'},
```
Comment (jgedeon-osi) [inline spec/javascripts/reducers/contactFormReducerSpec.js]: Am I missing something or should legacy_descriptor here be an object?

File: spec/javascripts/reducers/contactFormReducerSpec.js [hunk 8]
```diff
@@ -48,6 +54,23 @@ describe('contactReducer', () => {
...
             value: 'date time',
           },
+          people: [
+            {
+              first_name: 'Bob',
+              last_name: 'Smith',
+              middle_name: undefined,
+              name_suffix: undefined,
+              selected: false,
+              legacy_descriptor: '1',
```
Comment (jgedeon-osi) [inline spec/javascripts/reducers/contactFormReducerSpec.js]: legacy_descriptor object? here too?

File: spec/javascripts/selectors/contactFormSelectorsSpec.js [hunk 9]
```diff
@@ -1,122 +1,197 @@
...
-        fieldA: 1,
-        fieldB: undefined,
-        fieldC: 2,
-      })
-    )
+  describe('getFormattedContactPeople', () => {
+    it('returns the list of people on a contact form', () => {
+      const contactForm = {
+        people: [
+          {first_name: 'Bob', last_name: 'Smith', selected: true, legacy_descriptor: '1'},
```
Comment (jgedeon-osi) [inline spec/javascripts/selectors/contactFormSelectorsSpec.js]: legacy_descriptor object? here too?

File: spec/javascripts/reducers/contactFormReducerSpec.js (hunk 7 above)
Comment (saraOSI) [reply spec/javascripts/reducers/contactFormReducerSpec.js]: From the API it is indeed an object. For the purposes of the tests, I didn't think it mattered.

**Pull request merged**